src/cargo/util/toml.rs

```diff
@@ -505,7 +525,24 @@ fn inferred_bin_targets(name: &str, layout: &Layout) -> Vec<TomlTarget> {
                       *bin == layout.root.join("src").join("main.rs") {
             Some(name.to_string())
         } else {
-            bin.file_stem().and_then(|s| s.to_str()).map(|f| f.to_string())
+            // bin is either a source file or a directory with main.rs inside.
+            if bin.ends_with("main.rs") {
```

Comment:
Let's change this to `bin.ends_with("main.rs") && !bin.ends_with("src/bin/main.rs")` so that we one `if` inside.